power_grid_model_c/power_grid_model/include/power_grid_model/main_model.hpp

```diff
@@ -162,10 +162,18 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
         assert(construction_complete_);
 
         if constexpr (CacheType::value) {
+            constexpr auto comp_index = AllComponents::template index_of<CompType>();
+
+            auto& cached_sequence = std::get<comp_index>(cached_updated_sequence_indices_);
+            if (sequence_idx.empty()) {
+                main_core::get_component_sequence<CompType>(state_, begin, end, std::back_inserter(cached_sequence));
+            } else {
+                assert(cached_sequence.empty());
+                cached_sequence = sequence_idx;
```

Comment:
i don't like this copy but it was the only way i could think of that does not involve significant code changes to the batch calculation implementation.

The issue lies in the fact that we basically calculate the same thing (a cache of the ids lookup as a sequence that is by definition the same for all ids in a scenario) twice for independent batch update datasets: once for the entire batch, and once for the restore components.